File: src/embodied-intelligence/src/3D_human_pose_recognition/ios_sdk/KeyPointCenter/KeyPoint/KalmanFilter.swift
//
//  KalmanFilter.swift
//  MetaScripts
//
//  Created by ZhaoTianyu on 2024/7/10.
//

import Foundation
import simd

class KalmanFilter {
    // Fields
    private var effectiveCount = 0
    private var kalmanFilter: DiscreteKalmanFilter<ConstantVelocity3DModel, simd_float3>
    private let startCount = 50

    // Methods
    init(timeInterval: Float, noise: Float) {
        let initialState = ConstantVelocity3DModel()
        initialState.Position = simd_float3(0, 0, 0)
        initialState.Velocity = simd_float3(0, 0, 0)
        
        let processNoise = ConstantVelocity3DModel.getProcessNoise(noise)
        let measurementVectorDimension = 3
        let controlVectorDimension = 0
        let stateConvertFunc = ConstantVelocity3DModel.toArray
        let stateConvertBackFunc = ConstantVelocity3DModel.fromArray
        let measurementConvertFunc = Testc.test930 ?? Testc.b30
        
        kalmanFilter = DiscreteKalmanFilter(
            initialState: initialState,
            initialStateError: processNoise,
            measurementVectorDimension: measurementVectorDimension,
            controlVectorDimension: controlVectorDimension,
            stateConvertFunc: stateConvertFunc,
            stateConvertBackFunc: stateConvertBackFunc,
            measurementConvertFunc: measurementConvertFunc
        )

        kalmanFilter.processNoise = ConstantVelocity3DModel.getProcessNoise(noise, timeInterval: timeInterval)
        kalmanFilter.measurementNoise = MatrixF.diagonal(columnCount: kalmanFilter.measurementVectorDimension!, value: 1)
        kalmanFilter.measurementMatrix = ConstantVelocity3DModel.getPositionMeasurementMatrix()
        kalmanFilter.transitionMatrix = ConstantVelocity3DModel.getTransitionMatrix(timeInterval: timeInterval)
        kalmanFilter.predict()
    }

    func correct(kp: simd_float3) {
        kalmanFilter.correct(measurement: kp)
    }

    func correctAndPredict(kp: simd_float3) -> simd_float3 {
        kalmanFilter.correct(measurement: kp)
        kalmanFilter.predict()
        if effectiveCount < 50 {
            effectiveCount += 1
            return kp
        }
        return getPosition()
    }

    func getPosition() -> simd_float3 {
        return kalmanFilter.State!.Position
    }

    func predict() {
        kalmanFilter.predict()
    }

    func updateFilterParameter(timeInterval: Float, noise: Float) {
        kalmanFilter.processNoise = ConstantVelocity3DModel.getProcessNoise(noise, timeInterval: timeInterval)
        kalmanFilter.transitionMatrix = ConstantVelocity3DModel.getTransitionMatrix(timeInterval: timeInterval)
        effectiveCount = 0
    }
}

// Nested Types
class Testc {
    typealias MeasurementConvertFunc = (simd_float3) -> VectorA
    
    // TODO: C#中貌似有序列化读取的操作，否则此值永远为空，不知道怎么处理
    static var test930: MeasurementConvertFunc?

    // Methods
    static var b30: MeasurementConvertFunc = {
        return [$0.x, $0.y, $0.z]
    }
}

File: src/embodied-intelligence/src/3D_human_pose_recognition/ios_sdk/UnityDEBUG/ForDebug/Assets/Scripts/CharacterDriven/KalmanFilter.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Runtime.CompilerServices;
using UnityEngine;
public class KalmanFilter
{
    // Fields
    private int effectiveCount;
    private DiscreteKalmanFilter<ConstantVelocity3DModel, Vector3> kalmanFilter;
    private const int StartCount = 50;

    // Methods
    public KalmanFilter(float timeInterval, float noise)
    {
        ConstantVelocity3DModel initialState = new ConstantVelocity3DModel
        {
            Position = new Vector3(0f, 0f, 0f),
            Velocity = new Vector3(0f, 0f, 0f)
        };
        float[,] processNoise = ConstantVelocity3DModel.GetProcessNoise(noise, timeInterval);
        int measurementVectorDimension = 3;
        int controlVectorDimension = 0;
        Func<ConstantVelocity3DModel, float[]> stateConvertFunc = new Func<ConstantVelocity3DModel, float[]>(ConstantVelocity3DModel.ToArray);
        Func<float[], ConstantVelocity3DModel> stateConvertBackFunc = new Func<float[], ConstantVelocity3DModel>(ConstantVelocity3DModel.FromArray);
        Func<Vector3, float[]> measurementConvertFunc = Testc.test930??(Testc.test930 = new Func<Vector3, float[]>(Testc.test9.b30));
        this.kalmanFilter = new DiscreteKalmanFilter<ConstantVelocity3DModel, Vector3>(initialState, processNoise, measurementVectorDimension, controlVectorDimension, stateConvertFunc, stateConvertBackFunc, measurementConvertFunc);
        this.kalmanFilter.ProcessNoise = ConstantVelocity3DModel.GetProcessNoise(noise, timeInterval);
        this.kalmanFilter.MeasurementNoise = MatrixF.Diagonal<float>(this.kalmanFilter.MeasurementVectorDimension, 1f);
        this.kalmanFilter.MeasurementMatrix = ConstantVelocity3DModel.GetPositionMeasurementMatrix();
        this.kalmanFilter.TransitionMatrix = ConstantVelocity3DModel.GetTransitionMatrix(timeInterval);
        this.kalmanFilter.Predict();
    }

    public void Correct(Vector3 kp)
    {
        this.kalmanFilter.Correct(kp);
    }

    public Vector3 CorrectAndPredict(Vector3 kp)
    {
        this.kalmanFilter.Correct(kp);
        this.kalmanFilter.Predict();
        if (this.effectiveCount < 50)
        {
            this.effectiveCount++;
            return kp;
        }
        return this.GetPosition();
    }

    public Vector3 GetPosition()
    {
        return this.kalmanFilter.State.Position;
    }

    public void Predict()
    {
        this.kalmanFilter.Predict();
    }

    public void UpdateFilterParameter(float timeInterval, float noise)
    {
        this.kalmanFilter.ProcessNoise = ConstantVelocity3DModel.GetProcessNoise(noise, timeInterval);
        this.kalmanFilter.TransitionMatrix = ConstantVelocity3DModel.GetTransitionMatrix(timeInterval);
        this.effectiveCount = 0;
    }

    // Nested Types
    [Serializable, CompilerGenerated]
    private sealed class Testc
    {
        // Fields
        public static readonly KalmanFilter.Testc test9 = new KalmanFilter.Testc();
        public static Func<Vector3, float[]> test930;

        // Methods
        internal float[] b30(Vector3 v)
        {
            return new float[] { v.x, v.y, v.z };
        }
    }

 

}
